Move D3D11 API usage out of dxbc_debug.cpp

skarolewics opened the pull request:
The following instructions used D3D11 calls to compute results, and have been
moved to an API wrapper class that is defined in d3d11_shaderdebug.cpp:
transcendentals (rcp, rsqrt, exp2, log2, and sincos), sample info/pos, bufinfo,
resinfo, and sample/load/gather/lod.

baldurk commented on renderdoc/driver/shaders/dxbc/dxbc_debug.h:
```diff
@@ -147,6 +147,31 @@ struct GlobalState
+  virtual bool CalculateExp(const ShaderVariable &input, ShaderVariable &output) = 0;
+  virtual bool CalculateLog(const ShaderVariable &input, ShaderVariable &output) = 0;
+  virtual bool CalculateSinCos(const ShaderVariable &input, ShaderVariable &output1,
+                               ShaderVariable &output2) = 0;
```
This seems like a good candidate to merge into a single `CalculateMath` function or something like that with the `DXBC::OpcodeType` as a parameter. In your D3D11 implementation it forwards onto a single function anyway with a different string for performing the operation itself, which could be replaced by a switch(). I would expect the same thing to happen for any other implementation so we might as well codify that in the interface and make it easier to extend if we decide we want to run other operations on the GPU.

(In the past I've run into driver/GPU specific behaviour around certain edge cases so I could see wanting to move other operations either entirely onto the GPU or doing little feature-detect type things to see how the GPU reacts to edge cases with denormals and NaN/Inf to adjust the emulation)

skarolewics replied:
Merged these into a single function.

baldurk commented on renderdoc/driver/shaders/dxbc/dxbc_debug.h:
```diff
@@ -147,6 +147,31 @@ struct GlobalState
+  virtual ShaderVariable GetSampleInfo(const DXBC::ASMOperation &op) = 0;
+  virtual ShaderVariable GetBufferInfo(const DXBC::ASMOperation &op) = 0;
+  virtual ShaderVariable GetResourceInfo(const DXBC::ASMOperation &op, uint32_t mipLevel,
+                                         int &dim) = 0;
```
I'd rather not 'leak' the whole `ASMOperation` to this external function, and have it be responsible for knowing which operand is needed from which opcode. From what I can see in these `Get*Info` functions the op is only used (aside from the error string) to determine the binding - UAV/SRV and binding number, or perhaps rasterizer in one case. I'd rather do that operation-specific decode in the debug code itself then call out through this interface just specifying the binding alone.

skarolewics replied:
To clarify, are you suggesting combining these Get*Info functions into a single call, similar to the math functions? I took a preliminary look, and it feels like the majority of work would be branched, so leaving it as separate functions makes more sense to me. I agree about distilling the params down to strictly what is needed, making that change now.

I left some of the work for these opcodes in dxbc_debug.cpp (such as handling sample patterns, swizzling, etc.) since I couldn't think of a reason why D3D11/D3D12 would need to diverge there. Would you prefer making the API wrapper handle all of it, but moving this common functionality to a shared location, outside of dxbc_debug.cpp?

baldurk replied:
I don't mean we should combine these into a single call - I agree with you that the implementations would mostly be separate anyway. What I mean is e.g. looking at `GetSampleInfo` it has this code:

```
  else if(op.operands[1].type == DXBC::TYPE_RESOURCE && op.operands[1].indices.size() == 1 &&
          op.operands[1].indices[0].absolute && !op.operands[1].indices[0].relative)
```

which is a bit of a leaky abstraction - now each implementation of `DebugAPIWrapper` needs to know not only how to map the binding and get the sample info data, but also needs to know that specifically operand 1 is the one to look at, not operand 0. Similarly in `GetResourceInfo` it needs to look up operand 2.

What I'm suggesting is that for these functions the small amount of code that parses/decodes the op should be kept in dxbc_debug.cpp and it should just pass through e.g. an enum for rasterizer/srv/uav and a binding number.

skarolewics replied:
Understood and fixed.

baldurk commented on renderdoc/driver/shaders/dxbc/dxbc_debug.h:
```diff
@@ -147,6 +147,31 @@ struct GlobalState
+  virtual bool CalculateSampleGather(const DXBC::ASMOperation &op,
+                                     const std::vector<ShaderVariable> &srcOpers,
+                                     ShaderVariable ddxCalc, ShaderVariable ddyCalc,
+                                     ShaderVariable &output) = 0;
```
This will be harder to split apart than the above `Get*Info` functions but I think it's worth doing. The debug engine should be responsible for decoding the opcode, extracting out the operands, and passing along the information about the type & binding of the resource and the texture operation needed. Then the implementation of this class should only be responsible for mapping that request to the actual API and generating/running/fetching the shader. At the moment these two classes are pretty coupled because a lot of the implementation of this function needs to then poke back into DXBC internals.

Ideally then the implementation of this class will have very little dependency on the DXBC namespace (aside from using some enums

skarolewics replied:
Hmm, I see what you mean, and I agree with the value of splitting this into several, more targeted functions for the DXBC code to call into. This will take a while to sort through. :)

baldurk replied:
Actually similarly to above in reverse, I'm not meaning to split the function apart. I think it's fine to be merged because there's a fair bit of commonality between the paths. Instead I'm referring to the same thing again - the implementation of CalculateSampleGather still has to do a lot of dxbc-specific work to decode the right operands, iterate over declarations to find resource bindings, etc.

Imagine if in future D3D11 was updated to accept SPIR-V as a shader encoding (hallelujah!), so we have a spirv_debug.cpp which wants to call the same interface but with a completely different bytecode. We want this interface to be mostly agnostic to the actual shader encoding it's debugging, it should simply be a provider for functions that we can't emulate on the CPU. We don't have to take it to extremes right now - it's fine to still use some DXBC enums like the opcode that could easily be genericised later - but I'd rather have the bulk of that dependency broken here.

skarolewics replied:
I see what you mean now, thanks for clarifying.

Taking another look, a lot of CalculateSampleGather is setting up all of the strings for the program (texture & sampler declaration, return type, texcoords, etc...it's a big list). That work feels best kept in DXBC since it'd be most likely just duplicated to D3D12. This would reduce code duplication but would result in a function signature full of std::strings which is extremely "go ahead and try not to shoot off your foot" feeling.

I'm inclined to make the function signature a soup of data from the ASMOperation that is needed, but add helper functions in DXBC to create each of those strings. Does that sound like the best strategy to you?

baldurk replied:
Having a 'derived' struct of just the configuration needed from ASMOperation is what I had in mind yes. I think it's fine to have the HLSL string generation in D3D11, even if it gets duplicated in D3D12. I'm not so much worried about a bit of duplication as I am with having this code separated but still entangled. I'd rather have all the D3D/HLSL stuff in the D3D11 project and all the DXBC stuff in the DXBC project first. Then if later we decide some code in the D3D11 codebase needs to be extracted out to share with D3D12 it may make sense to put that somewhere else, not in the DXBC project. There's already `renderdoc/driver/dx` - it's currently only used for the shared official headers but it could be used for storing common helpers.

I haven't gone through exactly what data would need to be passed, I think it'd be useful to codify that as a first step and see how much is actually left. It may boil down to not that much apart from the type/dimension of the resource and the type of the co-ordinates. If that leaves a couple of string lookup tables or helpers in D3D11 that's not a problem for now.

I'm not sure about this either because I have a look at it, but the D3D12 HLSL generation will have to be somewhat different anyway to handle the different binding model I think.

skarolewics replied:
Thanks. I didn't go as far as creating a struct for params for the SampleGather function, but despite the number of params, I think it's pretty sensible, and in doing this I found a bunch of code that could stay in DXBC.

baldurk reviewed (changes requested):
I think in principle this change is OK but the split of abstraction needs to be moved for the texture operations. The maths operations are nicely abstracted so there's no real dependency there - it takes an input and gives an output - and I'd like to see the texture operations organised the same way. That way rather than this being a move of code from one place to another, it's a better separation of concerns.

I haven't gone into finest detail over the rest of the code since it's a) largely identical and just moved and b) may change given the above so I'd be wasting your time. I'll re-review in more detail once the changes above are made.

baldurk commented on renderdoc/driver/shaders/dxbc/dxbc_debug.cpp:
```diff
@@ -2088,175 +2083,26 @@ State State::GetNext(GlobalState &global, State quad[4]) const
-      SAFE_RELEASE(prevUAV);
-
-      if(op.operation == OPCODE_SINCOS)
+      if(apiWrapper->CalculateMathIntrinsic(op.operation, srcOpers[0], calcResultA, calcResultB))
```
This will need to pass `srcOpers[1]` for `OPCODE_SINCOS` I think? Because `srcOpers` are fetched from `operands[1]` onwards without considering that this operation has two outputs so `operands[1]` is the second output.

skarolewics replied:
Great catch, fixed!

baldurk commented on renderdoc/driver/shaders/dxbc/dxbc_disassemble.h:
```diff
@@ -687,6 +687,8 @@ class DXBCFile;
 struct ASMIndex;
 struct ASMDecl;
 
+char *toString(OpcodeType op);
```
I'd prefer just to print the opcode as a number rather than making this public. The string isn't necessary for any error messages since it's not going to the user, it's indicating a coding error so anyone who sees it can look up the opcode by value.

skarolewics replied:
Good point, reverted this bit.

baldurk commented on renderdoc/driver/shaders/dxbc/dxbc_debug.h:
```diff
@@ -147,6 +147,33 @@ struct GlobalState
+                                     UINT samplerSlot, float samplerBias, const int texelOffsets[3],
+                                     int multisampleIndex, float lodOrCompareValue,
+                                     const char *swizzle, const char *gatherChannel,
+                                     const char *opString, ShaderVariable &output) = 0;
```
I tend not to be strict about this kind of thing but I think this function signature is a bit too large to have everything loose. It doesn't have to be all converted into a single struct, but at least you could gather texture parameters: `resourceDim`, `resourceRetType`, `resourceSampleCount`, `texSlot`, sampler parameters: `samplerMode`, `samplerSlot` and maybe also gather the sampling/co-ordinate parameters (though after combining the above it might be reasonable).

Also a slight nit-pick, but I'd prefer to pass the swizzle as `int[4]` with 0,1,2,3 values, and gatherChannel as an int with 0,1,2,3, then do the conversion to string on the other side. It's only a couple of lines/switch but I think it's better placed on the D3D side of things. The DXBC side can still convert 0xff for unused swizzles into 0 in the array.

skarolewics replied:
Bundled the resource and sampler data into single params for each, and moved the string creation of swizzle/channel to the D3D11 side.

baldurk commented:
I left a couple of notes about final tidying up but otherwise this looks good to me now.

CookiePLMonster commented on renderdoc/driver/shaders/dxbc/dxbc_debug.h:
```diff
@@ -147,6 +147,55 @@ struct GlobalState
   std::map<SampleEvalCacheKey, ShaderVariable> sampleEvalCache;
 };
 
+struct SampleGatherResourceData
```
It probably makes sense to put those structs inside `class DebugAPIWrapper` so they don't occupy global namespace?

skarolewics replied:
I did that initially, but the full name then felt very long, and I didn't want to remove 'SampleGather' from the struct name as that may cause someone to incorrectly assume that they have broader utility.

baldurk replied:
Yes, also it's already in `namespace ShaderDebug`. Even if it were in the global namespace I don't think that's a big deal.

baldurk commented on renderdoc/driver/shaders/dxbc/dxbc_debug.cpp:
```diff
@@ -4331,520 +3511,72 @@ State State::GetNext(GlobalState &global, State quad[4]) const
-          case 2: channel = "Blue"; break;
-          case 3: channel = "Alpha"; break;
-        }
+        gatherChannel = static_cast<GatherChannel>(op.operands[3].comps[0]);
```
Prefer a normal cast, not a C++ cast here.

baldurk reviewed (approved):
There's one minor warning in the CI to fix and a style change, otherwise this LGTM.

baldurk commented on renderdoc/driver/d3d11/d3d11_shaderdebug.cpp:
```diff
@@ -239,6 +240,1368 @@ static void FlattenVariables(const rdcarray<ShaderConstant> &constants,
+    if(retFmt == DXGI_FORMAT_UNKNOWN)
+    {
+      funcRet = buf;
+      retFmt = fmts[resourceData.dim];
```
Found a typo here while testing - this should be indexed by `resourceData.retType` not `resourceData.dim`.

skarolewics replied:
Oops, thanks for catching. Fixed.

skarolewics committed:
Move D3D11 API usage out of dxbc_debug.cpp

The following instructions used D3D11 calls to compute results, and have been
moved to an API wrapper class that is defined in d3d11_shaderdebug.cpp:
transcendentals (rcp, rsqrt, exp2, log2, and sincos), sample info/pos, bufinfo,
resinfo, and sample/load/gather/lod.
